File: Python/RagOrchestrator.py
from typing import List, Optional
import time
from .JsonCorpusLoader import load_corpus
from .RagContext import RagContext
from .TraceEvent import TraceEvent

class RagOrchestrator:
    def __init__(self, corpus_path, intent_detector, query_writer, retriever, reranker, answer_agent, trace_bus):
        self.corpus = load_corpus(corpus_path)
        self.intent_detector = intent_detector
        self.query_writer = query_writer
        self.retriever = retriever
        self.reranker = reranker
        self.answer_agent = answer_agent
        self.trace_bus = trace_bus
        self.context = RagContext() 
        # Raporlama için metrikleri burada tutuyoruz
        self.last_latency_ms = 0
        self.last_success = False

    def run(self, question: str) -> None:
        start_time = time.perf_counter() # Hassas zamanlama başlangıcı
        
        self.context = RagContext()
        self.context.question = question

        # 1. Intent Detection
        self.context.intent = self.intent_detector.detect(question) 
        
        # 2. Query Generation
        self.context.terms = self.query_writer.write(question, self.context.intent)
        
        # 3. Retrieval
        if hasattr(self.retriever, 'retrieve'):
            self.context.hits = self.retriever.retrieve(self.context.terms, self.corpus) 
        else:
            self.context.hits = self.retriever(self.context.terms, self.corpus)
            
        # 4. Reranking
        self.context.finalHits = self.reranker.rerank(self.context.terms, self.context.hits)
        
        # 5. Answer Generation
        self.context.answerText = self.answer_agent.generate_answer(question, self.context.terms, self.context.finalHits)
        
        # Metrikleri kaydet
        end_time = time.perf_counter()
        self.last_latency_ms = int((end_time - start_time) * 1000)
        self.last_success = len(self.context.finalHits) > 0

        # Cevabı bas
        print(self.context.answerText)

        # 6. Trace Loglarını Kaydet
        if self.trace_bus:
            self.trace_bus.on_event(TraceEvent("FullPipeline", question, "Success", self.last_latency_ms))

File: Java/RagOrchestrator.java
package mu.cse.rag.clean;

import java.util.List;

public class RagOrchestrator {

    private final IntentDetector intentDetector;
    private final QueryWriter queryWriter;
    private final Retriever retriever;
    private final Reranker reranker;
    private final AnswerAgent answerAgent;
    private final List<Chunk> corpus;
    private final TraceBus traceBus;
    private final RagContext context = new RagContext();


    public RagOrchestrator(String corpusPath,
                       IntentDetector intentDetector,
                       QueryWriter queryWriter,
                       Retriever retriever,
                       Reranker reranker,
                       AnswerAgent answerAgent,
                       TraceBus traceBus) {

    this.corpus = JsonCorpusLoader.loadCorpus(corpusPath);
    this.intentDetector = intentDetector;
    this.queryWriter = queryWriter;
    this.retriever = retriever;
    this.reranker = reranker;
    this.answerAgent = answerAgent;
    this.traceBus = traceBus;
}


    public void run(String question) {

    context.question = question;

    // 1) Intent detection
    long t1 = System.currentTimeMillis();
    context.intent = intentDetector.detect(question);
    long t2 = System.currentTimeMillis();
    traceBus.publish(new TraceEvent(
            "IntentDetection",
            question,
            context.intent.toString(),
            t2 - t1
    ));

    // 2) Query generation
    long t3 = System.currentTimeMillis();
    context.terms = queryWriter.write(question, context.intent);
    long t4 = System.currentTimeMillis();
    traceBus.publish(new TraceEvent(
            "QueryGeneration",
            question,
            context.terms.toString(),
            t4 - t3
    ));

    // 3) Retrieval
    long t5 = System.currentTimeMillis();
    context.hits = retriever.retrieve(context.terms, corpus);
    long t6 = System.currentTimeMillis();
    traceBus.publish(new TraceEvent(
            "Retrieval",
            context.terms.toString(),
            (context.hits == null ? "0 hits" : context.hits.size() + " hits"),
            t6 - t5
    ));

    // 4) Reranking
    long t7 = System.currentTimeMillis();
    context.finalHits = reranker.rerank(context.terms, context.hits);
    long t8 = System.currentTimeMillis();
    traceBus.publish(new TraceEvent(
            "Reranking",
            context.hits.toString(),
            context.finalHits.toString(),
            t8 - t7
    ));

    // 5) Answer generation
    long t9 = System.currentTimeMillis();
    context.answerText = answerAgent.generateAnswer(question, context.terms, context.finalHits);
    long t10 = System.currentTimeMillis();
    traceBus.publish(new TraceEvent(
            "AnswerGeneration",
            question,
            context.answerText,
            t10 - t9
    ));

    // 6) Show output
    System.out.println(context.answerText);
}

}
